### PR TITLE
Add COMMENT feature to ModuleMate Finder

### PR DESCRIPTION
**Pull request by Btaykb**
Implement `comment`  for ModuleMate Finder. 

### Adding a comment to a contact
- Similar to the tutorial on adding a remark, except that when the user inputs an empty value, the comment does not get deleted, an error is shown instead
- More intuitive for users to associate removing a comment with a separate command then simply inputting an empty string (To be implemented in the proposed `delete` command changes). Also to maintain consistency with the way ModuleMateFinder handles the deletion of `Person` and `Modules`
- For now, the deletion is handled by just inputting an empty string eg. `comment 1` will clear the comment of the 1st person (Temporary)

User Guide Excerpt 

Format: `comment INDEX c/COMMENT`

* Adds a comment for the person at the specified `INDEX`.
* `INDEX` must be a **positive integer** 1, 2, 3, ...
* `COMMENT` cannot be blank.
* Any existing comments for a person will be overwritten by the new user input.

Examples:
* `comment 2 c/Good at math.` will add the comment `Good at math` to the 2nd person.

### CODE REVIEW

**Comment by bakano98** [comment]
> ### 2) Removing a comment of a contact
> * Removes a comment with a designated command instead of leaving `addcomment` to handle both adding and deletion of comments
> 

Is there any justification for this though? Seems a bit overkill to have 2 commands when you could simply do have an empty addcomment, (e.g. `addcomment 1`) to remove the comment, similar to what `Status` is doing. Plus, doing it that way would reduce the amount of duplicate code

**Comment by Btaykb** [comment]
> > ### 2) Removing a comment of a contact
> > 
> > * Removes a comment with a designated command instead of leaving `addcomment` to handle both adding and deletion of comments
> 
> Is there any justification for this though? Seems a bit overkill to have 2 commands when you could simply do have an empty addcomment, (e.g. `addcomment 1`) to remove the comment, similar to what `Status` is doing. Plus, doing it that way would reduce the amount of duplicate code

I agree on the point about duplicate code. 

But with regards to `clearcomment`, I personally feel its a bit unintuitive that adding an empty comment will equates to removing the comment for a person which is why I extended it to a separate command. Similar to how `addmodule` and `clearmodules` work. 

Maybe because of the name of the command having an add in it. Can change to just `comment`

**Comment by bakano98** [comment]
> Maybe because of the name of the command having an add in it. Can change to just `comment`

Yep, could we do this? I think it would be better and reduces the amount of duplicate code

File: src/main/java/seedu/address/logic/commands/AddCommentCommand.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Add a comment to the person identified "
+            + "by the index number used in the displayed person list. "
+            + "Existing comments will be overwritten by the input.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + PREFIX_COMMENT + "COMMENT\n"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_COMMENT + "Good at teamwork and programming";
+    public static final String MESSAGE_ADD_COMMENT_SUCCESS = "Added remark to Person: %1$s";
+    public static final String MESSAGE_DELETE_COMMENT_SUCCESS = "Removed remark from Person: %1$s";
+
```
Comment (lawwm) [inline src/main/java/seedu/address/logic/commands/AddCommentCommand.java:32]: Change to comment.

**Review by lawwm — COMMENTED** [review]
LGTM

**Comment by fantablack** [comment]
Nice work for handling this new field! I have tested it with valid inputs and as well as the more common invalid inputs, and it seems to be working well.

With regards to the additional clearcomment command, I'm on the side of retaining the clearcommand in the name of internal consistency, as that is how our contacts and modules perform, with the exception of the status command.

**Review by fantablack — COMMENTED** [review]
Nice work for handling this new field! I have tested it with valid inputs and as well as the more common invalid inputs, and it seems to be working well.

With regards to the additional `clearcomment` command, I'm on the side of retaining the `clearcommand` in the name of internal consistency, as that is how our contacts and modules perform, with the exception of the status command.

Perhaps we can hear from the rest @lawwm  @bakano98 @Btaykb @junlee1991 on their thoughts on how we should standardize user interaction for our commands? Should we have an explicit add and remove command (like Person, Modules)? Or a toggle (like Status)

**Comment by Btaykb** [comment]
> Nice work for handling this new field! I have tested it with valid inputs and as well as the more common invalid inputs, and it seems to be working well.
> 
> With regards to the additional `clearcomment` command, I'm on the side of retaining the `clearcommand` in the name of internal consistency, as that is how our contacts and modules perform, with the exception of the status command.
> 
> Perhaps we can hear from the rest @lawwm @bakano98 @Btaykb @junlee1991 on their thoughts on how we should standardize user interaction for our commands? Should we have an explicit add and remove command (like Person, Modules)? Or a toggle (like Status)

I think both @bakano98 and @fantablack bring up valid points, maybe one way to maintain consistency with having a separate delete command (eg. `clearcomment`, `clearmodules`) while ensuring that not too much code duplication happens is to combine the `clear` commands. 

So for instance `clear 1 c/` will clear the comments for 1st person and `clear 1 m/` will clear the modules for first person etc. `clear` will still clear the address book as expected. That way all the clearing of specific fields logic is kept into one class (which can be easily modified) without having to keep creating a new class and result in excessive duplication. 

Maybe can hear from @lawwm, @fantablack, @bakano98, @junlee1991 on your thoughts.

File: src/main/java/seedu/address/logic/commands/AddCommentCommand.java (hunk 1 above)
Comment (Btaykb) [reply src/main/java/seedu/address/logic/commands/AddCommentCommand.java:32]: Good catch, will make the changes

**Comment by fantablack** [comment]
> > Nice work for handling this new field! I have tested it with valid inputs and as well as the more common invalid inputs, and it seems to be working well.
> > With regards to the additional `clearcomment` command, I'm on the side of retaining the `clearcommand` in the name of internal consistency, as that is how our contacts and modules perform, with the exception of the status command.
> > Perhaps we can hear from the rest @lawwm @bakano98 @Btaykb @junlee1991 on their thoughts on how we should standardize user interaction for our commands? Should we have an explicit add and remove command (like Person, Modules)? Or a toggle (like Status)
> 
> I think both @bakano98 and @fantablack bring up valid points, maybe one way to maintain consistency with having a separate clear command (eg. `clearcomment`, `clearmodules`) while ensuring that not too much code duplication happens is to combine the `clear` commands.
> 
> So for instance `clear 1 c/` will clear the comments for 1st person and `clear 1 m/` will clear the modules for first person etc. `clear` will still clear the address book as expected. That way all the clearing of specific fields logic is kept into one class (which can be easily modified) without having to keep creating a new class and result in excessive duplication.
> 
> Maybe can hear from @lawwm, @fantablack, @bakano98, @junlee1991 on your thoughts.

@Btaykb I understand your point of view. However I am concerned that the additional flags needed for the User to remember (c is for comments, m is for modules etc) will be too taxing on them. Moreover, standardizing our commands these way would also require Delete command to behave this way, adding on to the confusion that Delete and Clear might have if they both operate using flags.

**Comment by Btaykb** [comment]
PR changes:

- Removed the `clearcomment` command
- Rename `addcomment` to just simply `comment`
- Add test cases for CommentCommand and CommentCommandParser classes
- Clearing of `comment` is done via an empty comment command for now (eg. `comment  1`)


Seeing as to how there are no issues with the `comment` field, and the issues are mostly with the way `comment` should be deleted via the new `delete` command changes, will go ahead to merge this PR first to meet our V1.3 deadline on 23 March.

**Pull request merged**